internal/provider_framework/attribute_validators.go

```diff
@@ -0,0 +1,368 @@
+}
+
+var _ tfsdk.AttributeValidator = (*exactlyOneOfAttributeValidator)(nil)
+
+func (av exactlyOneOfAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av exactlyOneOfAttributeValidator) MarkdownDescription(_ context.Context) string {
+	return fmt.Sprintf("Ensure that one and only one attribute from this collection is set: %q", av.attrPaths)
+}
+
+func (av exactlyOneOfAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	// Assemble a slice of paths, ensuring we don't repeat the attribute this validator is applied to
+	var paths []*tftypes.AttributePath
+	if containsAttrPath(req.AttributePath, av.attrPaths...) {
+		paths = av.attrPaths
+	} else {
+		paths = append(av.attrPaths, req.AttributePath)
+	}
+
+	tflog.Debug(ctx, "Validating that one and only one attribute is set", map[string]interface{}{
+		"exactlyOneOfAttributes": paths,
+	})
+
+	count := 0
+	for _, path := range paths {
+		var v attr.Value
+		req.Config.GetAttribute(ctx, path, &v)
+
+		if !v.IsNull() {
+			count++
+		}
+	}
+
+	if count == 0 {
+		res.Diagnostics.AddAttributeError(
+			req.AttributePath,
+			"Invalid combination of arguments: no attribute set, when one and only one was expected",
+			fmt.Sprintf("No attribute out of %q has been set", joinAttrPathsToString(paths...)),
+		)
+	}
+
+	if count > 1 {
+		res.Diagnostics.AddAttributeError(
+			req.AttributePath,
+			"Invalid combination of arguments: more than one attribute set, when only one was expected",
+			fmt.Sprintf("More than one attribute out of %q has been set", joinAttrPathsToString(paths...)),
+		)
+	}
+}
+
+// atLeastOneOfAttributeValidator checks that of a set of *tftypes.AttributePath,
+//// including the attribute it's applied to, at least one attribute out of all specified is configured.
+////
+//// The provided tftypes.AttributePath must be "absolute",
+//// and starting with top level attribute names.
+type atLeastOneOfAttributeValidator struct {
+	attrPaths []*tftypes.AttributePath
+}
+
+// atLeastOneOf is a helper to instantiate exactlyOneOfAttributeValidator.
+func atLeastOneOf(attributePaths ...*tftypes.AttributePath) tfsdk.AttributeValidator {
+	return &atLeastOneOfAttributeValidator{attributePaths}
+}
+
+var _ tfsdk.AttributeValidator = (*atLeastOneOfAttributeValidator)(nil)
+
+func (av atLeastOneOfAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av atLeastOneOfAttributeValidator) MarkdownDescription(ctx context.Context) string {
+	return fmt.Sprintf("Ensure that at least one attribute from this collection is set: %q", av.attrPaths)
+}
+
+func (av atLeastOneOfAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	// Assemble a slice of paths, ensuring we don't repeat the attribute this validator is applied to
```

Comment:
Is a check for `if req.AttributeConfig.IsNull() || req.AttributeConfig.IsUnknown()` required here?
If not, should there be one for https://github.com/hashicorp/terraform-provider-tls/pull/215/files#diff-c32022ba2194630d49dde2e8e91220e983a86076686d51bbc96f3edc6aba6f29R117 ?

Reply:
Well, for those validators their whole logic is based around their `null-ness`. So they will be checked for null during the execution.

The other validators, where I'm checking for null/unknown, their actual value is validated.